### PR TITLE
Batch table

### PR DESCRIPTION
**Pull request by bagnell**
Adds a batch table similar to what is used by 3D tiles, but can also be used for batched polylines and geometry. Currently only hooked up to the `PolylineCollection`. It can replace the implementation of per-instance attributes for `Primitive`, but that won't happen for this PR.

This needs to be merged before #4309.

### CODE REVIEW

**Comment by pjcozzi** [comment]
> It can replace the implementation of per-instance attributes for Primitive, but that won't happen for this PR.

Submit an issue or update the appropriate roadmap.

File: Source/Scene/BatchTable.js [hunk 1]
```diff
@@ -0,0 +1,475 @@
...
+     * @private
+     *
+     * @param {Object[]} attributes An array of objects describing a per instance attribute. Each object contains a datatype, components per attributes, whether it is normalized and a function name
+     *     to retrieve the value in the vertex shader.
+     * @param {Number} numberOfInstances The number of instances in a batch table.
+     *
+     * @example
+     * // create the batch table
+     * var attributes = [{
+     *     functionName : 'getShow()',
```
Comment (pjcozzi) [inline Source/Scene/BatchTable.js]: `()` here, but not with `getPickColor`.  Which is it?

File: Source/Scene/BatchTable.js [hunk 2]
```diff
@@ -0,0 +1,475 @@
...
+     * attribute float batchId;
+     *
+     * void main() {
+     *     // ...
+     *     float show = getShow(batchId);
+     *     vec3 pickColor = getPickColor(batchId);
+     *     // ...
+     * }
+     */
+    function BatchTable(attributes, numberOfInstances) {
```
Comment (pjcozzi) [inline Source/Scene/BatchTable.js]: This can be `@private`, right?

File: Source/Scene/BatchTable.js [hunk 3]
```diff
@@ -0,0 +1,475 @@
...
+    /**
+     * Creates/updates the batch table texture.
+     * @param {FrameState} frameState The frame state.
+     *
+     * @exception {RuntimeError} The floating point texture extension is required but not supported.
+     */
+    BatchTable.prototype.update = function(frameState) {
+        var context = frameState.context;
+        if (this._pixelDatatype === PixelDatatype.FLOAT && !context.floatingPointTexture) {
+            // TODO: We could probably pack the floats to RGBA unsigned bytes but that would add a lot CPU and memory overhead.
```
Comment (pjcozzi) [inline Source/Scene/BatchTable.js]: Leave the comment, but drop the `TODO`; I don't expect that we would implement this fallback given how rare it should be.

File: Source/Scene/BatchTable.js (hunk 2 above)
Comment (pjcozzi) [inline Source/Scene/BatchTable.js]: Although a good bit of this code is clearly from the 3D Tiles batch table, it doesn't look realistic that 3D Tiles and Cesium primitives could share a common class.  Do you agree?

If so, is there significant helper functions that are worth sharing?

File: Source/Scene/BatchTable.js [hunk 4]
```diff
@@ -0,0 +1,475 @@
...
+     *
+     * @param {Number} instanceIndex The index of the instance.
+     * @param {Number} attributeIndex The index of the attribute.
+     * @param {undefined|Cartesian2|Cartesian3|Cartesian4} [result] The object onto which to store the result. The type is dependent on the attribute's number of components.
+     * @returns {Number|Cartesian2|Cartesian3|Cartesian4} The attribute value stored for the instance.
+     *
+     * @exception {DeveloperError} instanceIndex is out of range.
+     * @exception {DeveloperError} attributeIndex is out of range.
+     */
+    BatchTable.prototype.getEntry = function(instanceIndex, attributeIndex, result) {
```
Comment (pjcozzi) [inline Source/Scene/BatchTable.js]: Throughout is `Entry` the best term?  Seems like we just introduced it.  What is this called in Cesium primitives, an attribute, right?  Perhaps `getBatchedAttribute` and friends?

File: Source/Scene/PolylineCollection.js [hunk 5]
```diff
@@ -1212,22 +1236,30 @@ define([
...
+                    texCoordExpandAndBatchIndexArray[texCoordExpandAndBatchIndexIndex + 3] = polylineBatchIndex;
 
                     positionIndex += 6 * 3;
-                    colorIndex += 4;
-                    texCoordExpandWidthAndShowIndex += 4;
+                    texCoordExpandAndBatchIndexIndex += 4;
                 }
             }
+
+            // TODO: add support for color from the batch table?
```
Comment (pjcozzi) [inline Source/Scene/PolylineCollection.js]: What's the plan?

File: Source/Scene/BatchTable.js (hunk 2 above)
Comment (bagnell) [inline Source/Scene/BatchTable.js]: Yes, it is `@private`.

I don't see why 3D Tiles wouldn't use this as a helper class. Unless you want to keep the option of not having to require VTF, then we could separate this out into helper functions.

**Comment by bagnell** [comment]
@pjcozzi This is ready for another look.

File: Source/Scene/BatchTable.js [hunk 6]
```diff
@@ -0,0 +1,475 @@
...
+            }
+        }
+    });
+
+    function getDatatype(attributes) {
+        var foundFloatDatatype = false;
+        var length = attributes.length;
+        for (var i = 0; i < length; ++i) {
+            if (attributes[i].componentDatatype !== ComponentDatatype.UNSIGNED_BYTE) {
+                foundFloatDatatype = true;
```
Comment (pjcozzi) [inline Source/Scene/BatchTable.js]: Add a `break`?

**Comment by pjcozzi** [comment]
Even though this doesn't change the Cesium API, can you update CHANGES.md to say that the extension is now required?

**Comment by pjcozzi** [comment]
The following test passes in master and fails in this branch:

```
Scene/PolylineCollection does not render with width 0.0
Expected [ 0, 0, 0, 255 ] not to equal [ 0, 0, 0, 255 ].
```

**Comment by pjcozzi** [comment]
Just those comments.

**Comment by bagnell** [comment]
@pjcozzi This is ready.

**Pull request merged**